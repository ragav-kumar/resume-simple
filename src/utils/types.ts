export const endpoints = {
	/** Represents all site content that hails from the DB. Might as well load this in one go */
	content: "content",
	/** For PDF Generation */
	pdf: "pdf",
	/** For Contact form submissions */
	contact: "contact",
} as const;

export type Endpoints = keyof typeof endpoints;

export const routes = {
	profile: "/",
	competency: "/competencies",
	education: "/education",
	contact: "/contact",
	history: "/experience",
	pdf: "/pdf",
} as const;

/**
 * No parameters for the content request.
 * Only reason it's not pre-loaded is because of the fancy loading spinner
 */
export type ContentRequest = undefined;
export interface ContentResponse {
	profile: string[];
	competencies: Competency[];
	education: Education[];
	experience: Experience[];
}

export interface CompetencyItem {
	label: string|null;
	text: string;
}
export interface Competency {
	name: string;
	items: CompetencyItem[];
}

export interface Education {
	title: string;
	location: string;
	start: string;
	end: string|null;
	dateOverride: string|null;
}
export interface Experience {
	title: string;
	employer: string;
	location: string;
	start: string;
	end: string|null;
	dateOverride: string|null;
}
