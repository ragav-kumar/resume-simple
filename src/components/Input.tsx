import React from "react";
import { useField, useFormikContext } from "formik";
import styled, { css } from "styled-components";

interface BaseInputProps {
	name: string;
	// Typescript complains when matching the ref types
	ref?: any;
}

type InputProps = BaseInputProps & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLInputElement>, HTMLInputElement>;
type TextareaProps = BaseInputProps & React.DetailedHTMLProps<React.InputHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement>;

/**
 * Formik Text input field
 */
export const Input = ( props: InputProps ) => {
	const { submitCount } = useFormikContext<{ [field: string]: string; }>();
	const [ field, { error, touched } ] = useField<string>(props);
	const isError = !!error && touched && submitCount > 0;

	return (
		<StyledInput
			{...field}
			{...props}
			isError={isError}
			data-testid={props.name}
		/>
	);
}

export const Textarea = ( props: TextareaProps ) => {
	const { submitCount } = useFormikContext<{ [field: string]: string; }>();
	const [ field, { error, touched } ] = useField<string>(props);
	const isError = !!error && touched && submitCount > 0;

	return (
		<StyledTextarea
			{...field}
			{...props}
			isError={isError}
			data-testid={props.name}
		/>
	);
}

interface StyledInputProps {
	isError: boolean;
}

const inputCss = ( { isError }:StyledInputProps) => css`
	border-color: ${isError ? "red" : "#767676"};
	font-size: 20px;
	padding: .2em;
	margin: .2em 0;
	border-style: solid;
	border-width: 1px;
	box-shadow: none;
`;

const StyledInput = styled.input<StyledInputProps>`
	${inputCss};
`;

const StyledTextarea = styled.textarea<StyledInputProps>`
	${inputCss};
	height: 4em;
	resize: vertical;
`;
