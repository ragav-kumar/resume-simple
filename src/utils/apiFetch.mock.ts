import { ContentResponse, Endpoints } from "./types";
import reciprocity from "../img/reciprocity.png";
import cityautobahn from "../img/cityautobahn.png";
import acc from "../img/accredited-accommodations.png";

export const apiFetch = async <
	Request extends unknown,
	Response extends unknown
	>( endpoint: Endpoints, _data?: Request ):Promise<Response> => {

	switch (endpoint ) {
		case "pdf":
			return Promise.resolve(process.env.REACT_APP_API_URL! + "/resume.pdf" as Response);
		case "contact":
			return Promise.resolve(true as any as Response);
		case "content":
			return Promise.resolve(await fetchMockContent() as Response);
	}
	return Promise.reject("");
}

const fetchMockContent = async ():Promise<ContentResponse> => {
	const fetchResponse = await fetch(process.env.REACT_APP_API_URL! + "/content.json", {
		method: 'GET',
		headers: {
			'Content-Type': 'application/json',
			'Access-Control-Allow-Origin': "*",
		},
	});
	const json = await fetchResponse.json() as ContentResponse;

	json.portfolio.delivered.forEach(( { imgUrl }, index) => {
		switch (imgUrl ) {
			case "reciprocity":
				json.portfolio.delivered[index].imgUrl = reciprocity;
				break;
			case "cityautobahn":
				json.portfolio.delivered[index].imgUrl = cityautobahn;
				break;
			case "acc":
				json.portfolio.delivered[index].imgUrl = acc;
				break;
		}
	})

	return json;
}
