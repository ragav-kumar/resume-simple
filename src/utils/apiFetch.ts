import { endpoints, Endpoints } from "./types";

const clientId = process.env.REACT_APP_CLIENT_ID || "";
const apiUrl = process.env.REACT_APP_API_URL || "http://localhost/";

interface FetchResponseSuccess<T extends unknown> {
	success: true;
	data: T;
	error: undefined;
}

interface FetchResponseFailure {
	success: false;
	data: undefined;
	error: string;
}

export const apiFetch = async <
	Request extends unknown,
	Response extends unknown
	>( endpoint: Endpoints, data?: Request ):Promise<Response> => {
	try {
		const body = new URLSearchParams({
			client_id: clientId,
			data: JSON.stringify(data),
		}).toString();

		const response = await fetch(apiUrl + endpoints[endpoint], {
			method: 'POST',
			mode: 'no-cors',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded',
				'Access-Control-Allow-Origin': '*',
			},
			body: body,
		});

		const json = (await response.json()) as FetchResponseSuccess<Response> | FetchResponseFailure;
		return json.success ? Promise.resolve(json.data) : Promise.reject(json.error);
	} catch ( e ) {
		console.log(e)
		return Promise.reject("Connection Error");
	}
}
