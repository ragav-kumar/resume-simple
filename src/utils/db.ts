import { createContext } from 'react';
import { ContentResponse } from "./types";
/**
 * Mockup of DB data. Replace with a link to a mySQL DB later.
 */

/**
 * All data for a given section
 */
export type SectionData = {
	menuLabel: string;
	header   : string;
	path     : string;
};
/**
 * Menu Sections. Defined here cause it's used in multiple places
 */
export const sections:SectionData[] = [
	{
		menuLabel: "Profile",
		header: "Professional Profile",
		path: "/",
	}, {
		menuLabel: "Competencies",
		header: "Professional Competencies",
		path: "/competencies",
	}, {
		menuLabel: "Education",
		header: "Education",
		path: "/education",
	}, {
		menuLabel: "Experience",
		header: "Employment History",
		path: "/experience",
	},
];

export type DbContextType = ContentResponse;

export const blankContext:DbContextType = {
	profile: [],
	competencies: [],
	education: [],
	experience: [],
}
export const DbContext = createContext<DbContextType>(blankContext);