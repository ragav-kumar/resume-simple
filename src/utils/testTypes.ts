import { FieldHelperProps, FieldHookConfig, FieldInputProps, FieldMetaProps } from "formik";

export type FormikUseFieldArgType = string | FieldHookConfig<unknown>;
export type FormikUseFieldReturnType = [FieldInputProps<unknown>, FieldMetaProps<unknown>, FieldHelperProps<unknown>];
