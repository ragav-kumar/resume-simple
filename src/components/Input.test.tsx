import React from "react";
import renderer from 'react-test-renderer';
import { Input } from "./Input";
import { FormikContextType, useField, useFormikContext } from "formik";
import { mocked } from "ts-jest/utils";
import { FormikUseFieldArgType, FormikUseFieldReturnType } from "../utils/testTypes";

jest.mock('formik', () => ({
	...jest.requireActual<any>('formik'),
	useFormikContext: jest.fn(),
	useField: jest.fn()
}));

describe('<Input />', () => {
	it('should render correctly', () => {
		mocked(useFormikContext).mockReturnValue({ submitCount: 0 } as FormikContextType<unknown>);
		mocked(useField).mockImplementation((name:FormikUseFieldArgType) => [{
			name,
			value: "dummy",
		}, {
			error: undefined,
			touched: false,
		}, {}] as FormikUseFieldReturnType);
		const tree = renderer
			.create(<Input name="test-input"/>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it ('should change style on error', () => {
		mocked(useFormikContext).mockReturnValue({ submitCount: 1 } as FormikContextType<unknown>);
		mocked(useField).mockImplementation((name:FormikUseFieldArgType) => [{
			name,
			value: "dummy",
		}, {
			error: "test-error",
			touched: true,
		}, {}] as FormikUseFieldReturnType);
		const tree = renderer
			.create(<Input name="test-input"/>)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
