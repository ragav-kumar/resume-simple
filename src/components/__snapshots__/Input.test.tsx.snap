// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should change style on error 1`] = `
<input
  className="sc-bdfBQB dlwpfJ"
  name="test-input"
  value="dummy"
/>
`;

exports[`<Input /> should render correctly 1`] = `
<input
  className="sc-bdfBQB jXLNkU"
  name="test-input"
  value="dummy"
/>
`;
